src/modules/presence_reginfo/notify_body.c

```diff
@@ -0,0 +1,322 @@
+	xmlNewProp(root_node, BAD_CAST "state", BAD_CAST "full");
+
+	/* loop over all bodies and create the aggregated body */
+	for(i = 0; i < j; i++) {
+		// LM_DBG("[n]=%d, [i]=%d, [j]=%d xml_array[i]=%p\n", n, i, j,
+		// 		xml_array[j]);
+
+		// get the root reginfo element
+		p_root = xmlDocGetRootElement(xml_array[i]);
+		if(p_root == NULL) {
+			LM_ERR("the xml_tree root element is null\n");
+			goto error;
+		}
+
+		// get the children registration elements
+		if(p_root->children) {
+			// loop over registration elements
+			for(node = p_root->children; node; node = next_node) {
+				next_node = node->next;
+				if(node->type != XML_ELEMENT_NODE) {
+					continue;
+				}
+				LM_DBG("node type: Element, name: %s\n", node->name);
+
+				/* we do not copy the node, but unlink it and then add it ot the new node
+				 * this destroys the original document but we do not need it anyway.
+				 */
+				xmlUnlinkNode(node);
+				if(xmlAddChild(root_node, node) == NULL) {
+					xmlFreeNode(node);
+					LM_ERR("while adding child\n");
+					goto error;
+				}
+			} // end of loop over registration elements
+		}
+	} // end of loop over all bodies
+
+	// convert to string & cleanup
+	xml_array = (xmlDocPtr *)pkg_malloc(n * sizeof(xmlDocPtr));
+	if(xml_array == NULL) {
+		LM_ERR("while allocating memory");
+		return NULL;
+	}
+	memset(xml_array, 0, n * sizeof(xmlDocPtr));
+
+	body = (str *)pkg_malloc(sizeof(str));
+	if(body == NULL) {
+		ERR_MEM(PKG_MEM_STR);
+	}
+
+	xmlDocDumpFormatMemory(doc, (xmlChar **)(void *)&body->s, &body->len, 1);
+
+	for(i = 0; i < j; i++) {
+		if(xml_array[i] != NULL)
+			xmlFreeDoc(xml_array[i]);
+	}
+	if(doc) {
+		xmlFreeDoc(doc);
+	}
+	if(xml_array)
+		pkg_free(xml_array);
+
+	xmlCleanupParser();
+	xmlMemoryDump();
+
+	return body;
+
+	// error handling, cleanup and return NULL
+error:
+	if(xml_array != NULL) {
+		for(i = 0; i < j; i++) {
+			if(xml_array[i] != NULL)
+				xmlFreeDoc(xml_array[i]);
+		}
+		pkg_free(xml_array);
+	}
+	if(body) {
+		pkg_free(body);
+	}
+	if(doc) {
+		xmlFreeDoc(doc);
+	}
+
+	return NULL;
+}
+
+str *reginfo_body_setversion(subs_t *subs, str *body)
+{
+	char *version_start = 0;
+	char version[MAX_INT_LEN + 2]; /* +2 becasue of trailing " and \0 */
+	int version_len;
+	str *aux_body = NULL;
+
+	if(!body) {
+		return NULL;
+	}
+
+	/* xmlDocDumpFormatMemory creates \0 terminated string */
+	/* version parameters starts at minimum at character 30 */
+	if(body->len < 37) {
+		LM_ERR("body string too short!\n");
+		return NULL;
+	}
+	version_start = strstr(body->s + 30, "version=");
+	if(!version_start) {
+		LM_ERR("version string not found!\n");
+		return NULL;
+	}
+	version_start += 9;
+
+	/* safety check for placeholder - if it is body not set by the module,
+	 * don't update the version */
+	if(strncmp(version_start, "00000000000\"", 12) != 0)
+		return NULL;
+
+	version_len = snprintf(version, MAX_INT_LEN + 2, "%d\"", subs->version);
```

Comment:
you probably also want to check for negative return (encoding error)

Reply:
Well, does not change the flow but yes, is useful for debugging, added.